compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,564 @@
+      first: option(Location.t),
+    ) => {
+  let (_, startline, startc, _) = Locations.get_raw_pos_info(loc.loc_start);
+
+  let cmts = get_comments_on_line(startline, comments);
+  switch (cmts) {
+  | [] => cmts
+  | [fst, ...rem] =>
+    switch (first) {
+    | None => cmts
+    | Some(leading) =>
+      let fstclog = Locations.get_comment_loc(fst);
+
+      let (_, itemstartline, itemstartc, _) =
+        Locations.get_raw_pos_info(leading.loc_start);
+
+      if (itemstartline > startline) {
+        cmts;
+      } else {
+        let (_, cmtstartline, cmtstartc, _) =
+          Locations.get_raw_pos_info(fstclog.loc_start);
+        if (cmtstartline >= itemstartline && cmtstartc > itemstartc) {
+          [];
+        } else {
+          cmts;
+        };
+      };
+    }
+  };
+};
+
+let rec comments_inner =
+        (
+          prev: option(Parsetree.comment),
+          bracket_line_opt: option(int),
+          comments: list(Parsetree.comment),
+        ) => {
+  switch (prev) {
+  | None =>
+    switch (comments) {
+    | [] => [Doc.nil]
+    | [cmt, ...rem] =>
+      switch (bracket_line_opt) {
+      | None => [
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+      | Some(bracket_line) =>
+        let (_, this_line, _, _) =
+          Locations.get_raw_pos_info(
+            Locations.get_comment_loc(cmt).loc_start,
+          );
+
+        if (bracket_line == this_line) {
+          [
+            Doc.space,
+            comment_to_doc(cmt),
+            ...comments_inner(Some(cmt), bracket_line_opt, rem),
+          ];
+        } else {
+          [
+            Doc.hardLine,
+            comment_to_doc(cmt),
+            ...comments_inner(Some(cmt), bracket_line_opt, rem),
+          ];
+        };
+      }
+    }
+  | Some(prev_cmt) =>
+    switch (comments) {
+    | [] => [Doc.nil]
+    | [cmt, ...rem] =>
+      let (_, prev_line, _, _) =
+        Locations.get_raw_pos_info(
+          Locations.get_comment_loc(prev_cmt).loc_end,
+        );
+      let (_, this_line, _, _) =
+        Locations.get_raw_pos_info(Locations.get_comment_loc(cmt).loc_start);
+
+      switch (this_line - prev_line) {
+      | 0 => [
+          Doc.space,
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+
+      | 1 => [
+          switch (prev_cmt) {
+          | Line(_) => Doc.nil
+          | _ => Doc.hardLine
+          },
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+      | _ => [
+          switch (prev_cmt) {
+          | Doc(_)
+          | Shebang(_)
+          | Line(_) => Doc.nil
+          | _ => Doc.hardLine
+          },
+          Doc.hardLine,
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+      };
+    }
+  };
+};
+
+let inbetween_comments_to_docs =
+    (
+      ~offset: bool,
```

Comment:
I don't like this `~offset` argument. Why can't you just add the space in the calling code if their `offset` variable is true?

Reply:
Because we only insert if when there are comments, so we'd push the check for empty comments up to the callees, which seems to lead to duplicate code.